File: 528-swapping-nodes-in-a-linked-list/swapping-nodes-in-a-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapNodes(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        dummy = ListNode(0)
        dummy.next = head
        left = dummy
        count = k
        while count > 0:
            count -= 1
            left = left.next
        
        N = 0
        cur = head
        while cur:
            N += 1
            cur = cur.next

        right_idx = N - k
        right = head
        for i in range(N):
            if i == right_idx:
                break
            right = right.next
        
        left.val, right.val = right.val, left.val
        return dummy.next

        

        

File: 528-swapping-nodes-in-a-linked-list/swapping-nodes-in-a-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* swapNodes(ListNode* head, int k) {
        ListNode* left = head;
        
        for (int i = 0; i < k - 1; ++i) {
            left = left->next;
        }

        ListNode* cur = left;
        ListNode* right = head;
        while (cur->next) {
            cur = cur->next;
            right = right->next;
        }

        int tmp = left->val;
        left->val = right->val;
        right->val = tmp;

        return head;
    }
};